src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -82,18 +82,48 @@ internal static class DeserializationHelpers
             }
         }
 
-        internal static void ApplyInstanceAnnotations(object resource, IEdmStructuredTypeReference structuredType, ODataResource oDataResource,
+        internal static void ApplyInstanceAnnotations(object resource, IEdmStructuredTypeReference structuredType, ODataResourceBase oDataResource,
             ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
         {
-            PropertyInfo propertyInfo = EdmLibHelpers.GetInstanceAnnotationsContainer(structuredType.StructuredDefinition(), readContext.Model);
-            if (propertyInfo == null)
+            //Apply instance annotations for both entityobject/changedobject/delta and normal resources
+            IEntityObjectInstanceAnnotations entityObject = resource as IEntityObjectInstanceAnnotations;
+
+            if (entityObject != null)
             {
-                return;
+                SetInstanceAnnotationsForEdmChangedObject(entityObject, oDataResource, deserializerProvider, readContext);
             }
+            else
+            {
+                PropertyInfo propertyInfo = EdmLibHelpers.GetInstanceAnnotationsContainer(structuredType.StructuredDefinition(), readContext.Model);
+                if (propertyInfo == null)
+                {
+                    return;
+                }
 
-            IODataInstanceAnnotationContainer instanceAnnotationContainer = GetAnnotationContainer(propertyInfo, resource);
+                IODataInstanceAnnotationContainer instanceAnnotationContainer = GetAnnotationContainer(propertyInfo, resource);
 
-            SetInstanceAnnotations(oDataResource, instanceAnnotationContainer, deserializerProvider, readContext);
+                SetInstanceAnnotations(oDataResource, instanceAnnotationContainer, deserializerProvider, readContext);
+            }            
+        }
+
+        private static void SetInstanceAnnotationsForEdmChangedObject(IEntityObjectInstanceAnnotations entityObject, ODataResourceBase oDataResource, ODataDeserializerProvider deserializerProvider, 
+            ODataDeserializerContext readContext)
+        {
+            if (entityObject != null && oDataResource.InstanceAnnotations != null)
```

Comment:
>entityObject != null [](start = 16, length = 20)

You validated entityObject != null on line 91 before calling this method. Replace this with an assert.